plugins/steelseries/fu-steelseries-fizz-tunnel.c

```diff
@@ -0,0 +1,319 @@
+	    fu_device_get_version_format(device) == FWUPD_VERSION_FORMAT_UNKNOWN) {
+		g_autofree gchar *version = NULL;
+		version = fu_common_version_from_uint16(release, FWUPD_VERSION_FORMAT_BCD);
+		fu_device_set_version_format(device, FWUPD_VERSION_FORMAT_BCD);
+		fu_device_set_version(device, version);
+	}
+
+	/* add GUIDs in order of priority */
+	fu_device_add_instance_str(device, "STEELSERIES", "FIZZ");
+	fu_device_add_instance_u16(device, "VID", g_usb_device_get_vid(usb_device));
+	fu_device_add_instance_u16(device, "PID", g_usb_device_get_pid(usb_device));
+	fu_device_add_instance_u16(device, "REV", release);
+	fu_device_build_instance_id_quirk(device, NULL, "USB", "STEELSERIES", "VID", NULL);
+	fu_device_build_instance_id(device, NULL, "USB", "STEELSERIES", "VID", "PID", NULL);
+	fu_device_build_instance_id(device, NULL, "USB", "STEELSERIES", "VID", "PID", "REV", NULL);
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_steelseries_fizz_tunnel_setup(FuDevice *device, GError **error)
+{
+	FuDevice *parent = fu_device_get_parent(device);
+	guint32 calculated_crc;
+	guint32 stored_crc;
+	guint8 status;
+	guint8 fs = STEELSERIES_FIZZ_FILESYSTEM_MOUSE;
+	guint8 id = STEELSERIES_FIZZ_MOUSE_FILESYSTEM_BACKUP_APP_ID;
+	g_autofree gchar *version = NULL;
+
+	if (!fu_steelseries_fizz_connection_status(parent, &status, error))
+		return FALSE;
+	g_debug("ConnectionStatus: %u", status);
+	if (status == STEELSERIES_FIZZ_CONNECTION_STATUS_NOT_CONNECTED) {
+		fu_device_add_flag(device, FWUPD_DEVICE_FLAG_UNREACHABLE);
+
+		/* success */
+		return TRUE;
+	}
+
+	version = fu_steelseries_fizz_version(parent, TRUE, error);
+	if (version == NULL) {
+		g_prefix_error(error, "failed to get version: ");
+		return FALSE;
+	}
+	fu_device_set_version(device, version);
+
+	if (!fu_steelseries_fizz_file_crc32(parent,
+					    TRUE,
+					    fs,
+					    id,
+					    &calculated_crc,
+					    &stored_crc,
+					    error)) {
+		g_prefix_error(error,
+			       "failed to get file CRC32 from FS 0x%02x ID 0x%02x: ",
+			       fs,
+			       id);
+		return FALSE;
+	}
+
+	if (calculated_crc != stored_crc)
+		g_warning("%s: checksum mismatch, got 0x%08x, expected 0x%08x",
+			  fu_device_get_name(device),
+			  calculated_crc,
+			  stored_crc);
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_steelseries_fizz_tunnel_prepare(FuDevice *device,
+				   FuProgress *progress,
+				   FwupdInstallFlags flags,
+				   GError **error)
+{
+	guint8 status;
+	guint8 level;
+
+	if (!fu_steelseries_fizz_connection_status(fu_device_get_parent(device), &status, error)) {
+		g_prefix_error(error, "failed to get connection status: ");
+		return FALSE;
+	}
+	g_debug("ConnectionStatus: %u", status);
+	if (status == STEELSERIES_FIZZ_CONNECTION_STATUS_NOT_CONNECTED) {
+		g_set_error(error, FWUPD_ERROR, FWUPD_ERROR_NOT_FOUND, "device is unreachable");
+		return FALSE;
+	}
+	if (!fu_steelseries_fizz_battery_level(fu_device_get_parent(device), TRUE, &level, error)) {
```

Comment:
Hum, checking for battery level is too late at time of `_prepare()`.

Reply:
Is that a note for you or for me? :)

Reply:
for me, I have to do `_poll()` to inhibit update.

Reply:
[89b153e](https://github.com/fwupd/fwupd/pull/4670/commits/89b153e765cce644a0f044d7eea2d5f411ed462c)